InteractiveHtmlBom/web/render.js

```diff
@@ -480,21 +499,60 @@ function handleMouseUp(e, layerdict) {
   } else if (!redrawOnDrag) {
     redrawCanvas(layerdict);
   }
-  layerdict.transform.mousedown = false;
+
+  delete layerdict.pointerStates[e.pointerId];
 }
 
-function handleMouseMove(e, layerdict) {
-  if (!layerdict.transform.mousedown) {
+function handlePointerMove(e, layerdict) {
+  if (!layerdict.pointerStates.hasOwnProperty(e.pointerId)) {
     return;
   }
   e.preventDefault();
   e.stopPropagation();
-  var dx = e.offsetX - layerdict.transform.mousestartx;
-  var dy = e.offsetY - layerdict.transform.mousestarty;
-  layerdict.transform.panx += devicePixelRatio * dx / layerdict.transform.zoom;
-  layerdict.transform.pany += devicePixelRatio * dy / layerdict.transform.zoom;
-  layerdict.transform.mousestartx = e.offsetX;
-  layerdict.transform.mousestarty = e.offsetY;
+
+  if (!e.hasOwnProperty("offsetX")) {
+    // The polyfill doesn't set this properly
+    e.offsetX = e.pageX - e.currentTarget.offsetLeft;
+    e.offsetY = e.pageY - e.currentTarget.offsetTop;
+  }
+
+  var thisPtr = layerdict.pointerStates[e.pointerId];
+
+  if (Object.keys(layerdict.pointerStates).length == 1) {
+    // This is a simple drag
+    var dx = e.offsetX - thisPtr.lastX;
+    var dy = e.offsetY - thisPtr.lastY;
+
+    // If this number is low on pointer up, we count the action as a click
+    thisPtr.distanceTravelled += Math.abs(dx);
+    thisPtr.distanceTravelled += Math.abs(dy);
+
+    layerdict.transform.panx += devicePixelRatio * dx / layerdict.transform.zoom;
+    layerdict.transform.pany += devicePixelRatio * dy / layerdict.transform.zoom;
+  } else if (Object.keys(layerdict.pointerStates).length == 2) {
+    var otherPtr = layerdict.pointerStates[Object.keys(layerdict.pointerStates).filter((id) => id != e.pointerId)[0]];
+
+    // There's a multi-touch interaction happening, so neither pointer should be counted as doing a click
+    thisPtr.distanceTravelled = Infinity;
+    otherPtr.distanceTravelled = Infinity;
+
+    var oldDist = Math.sqrt(Math.pow(thisPtr.lastX - otherPtr.lastX, 2) + Math.pow(thisPtr.lastY - otherPtr.lastY, 2));
+    var newDist = Math.sqrt(Math.pow(e.offsetX - otherPtr.lastX, 2)     + Math.pow(e.offsetY - otherPtr.lastY, 2));
+
+    var scaleFactor = newDist/oldDist;
+
+    if (scaleFactor != NaN) {
+      layerdict.transform.zoom *= scaleFactor;
+
+      var zoomd = (1 - scaleFactor) / layerdict.transform.zoom;
+      layerdict.transform.panx += devicePixelRatio * otherPtr.lastX * zoomd;
+      layerdict.transform.pany += devicePixelRatio * otherPtr.lastY * zoomd;
```

Comment:
This effectively centers zooming action around otherPtr position. Doesn't it feel glitchy when you move both fingers during pinch/zoom and zooming center alternates between finger positions? Maybe that effect is not very noticeable, admittedly I have no easy way to try this out.

More natural alternative would be to center the zoom action on mid point between pointers.

Reply:
Thanks for the video explanation, consider this resolved.